flocker/volume/filesystems/openstack.py

```diff
@@ -0,0 +1,240 @@
+# Copyright Hybrid Logic Ltd.  See LICENSE file for details.
+
+"""
+ZFS APIs.
+"""
+
+from __future__ import absolute_import
+
+import os
+import string
+
+from subprocess import check_call
+import time
+
+from characteristic import with_cmp, with_repr
+
+from zope.interface import implementer
+
+from eliot import Logger
+
+from twisted.python.filepath import FilePath
+from twisted.internet.defer import succeed
+from twisted.application.service import Service
+
+from .interfaces import IFilesystem, IStoragePool
+
+from .._model import VolumeSize, VolumeName
+
+import pyrax
+
+import netifaces
+import ipaddr
+
+
+def get_public_ips():
+    ips = []
+    for interface in netifaces.interfaces():
+        interface_addresses = netifaces.ifaddresses(interface)
+        ipv4addresses = interface_addresses.get(netifaces.AF_INET, [])
+        for address in ipv4addresses:
+            ip = ipaddr.IPv4Address(address['addr'])
+            if not ip.is_private:
+                ips.append(ip)
+    return ips
+
+
+def driver_from_environment():
+    # Notes on how to get this working are in https://clusterhq.atlassian.net/browse/FLOC-1147
```

Comment:
This works for the spike, but let's discuss what's the best way to supply the cloud service credentials.